arangod/RestHandler/RestCollectionHandler.cpp

```diff
@@ -437,21 +438,24 @@ void RestCollectionHandler::handleCommandPut() {
             res = trx->begin();
 
             if (res.ok()) {
-              auto result = trx->truncate(coll->name(), opts);
+              OperationResult result = trx->count(coll->name(), transaction::CountType::Normal);
+              if (result.ok() && result.slice().isNumber()) {
+                count = result.slice().getNumber<uint64_t>();
+              }
+
+              result = trx->truncate(coll->name(), opts);
               res = trx->finish(result.result);
             }
           }
-
-          // note that previously we ran a compaction of the collection's key range
-          // and the collection's indexes key ranges directly after the truncate.
-          // this was done to make RocksDB physically free data of unused key ranges,
-          // but it seems that the used version of RocksDB is struggling to achieve
-          // that goal. although it does compact the range, it seems to read a lot
-          // of extra data from unrelated ranges as well, which blows up the overall
-          // time required to do the compactions. most recent versions of RocksDB
-          // do not seem to have this problem anymore. but for now, in this ArangoDb
-          // version we turn off the manual compaction and rely on RocksDB eventually
-          // freeing the now-unsed ranges.
+          // wait for the transaction to finish first. only after that compact the
+          // data range(s) for the collection
+          // we shouldn't run compact() as part of the transaction, because the compact
+          // will be useless inside due to the snapshot the transaction has taken
+          if (res.ok() && count >= 4 * 1024) {
+            // only compact if the collection contained a substantial amount of documents
+            // before truncation
+            coll->compact();
```

Comment:
Should this `compact` operation be asynchronous?

Reply:
Maybe. This basically restores the behavior that we had up to including 3.5.4, where we also ran the compact operation here, in the same way.
The diffence now is that we are only compacting if there have a few documents in the collection, so that truncating empty or almost empty collections does not lead to long, implausible delays.